sale_delivery_date/models/sale_order_line.py

```diff
@@ -43,129 +43,79 @@ def _prepare_procurement_values_commitment_date(self, res):
-        if calendar:
-            res["date_planned"] = calendar.plan_days(
-                -1, res["date_planned"], compute_leaves=True
-            )
-        # 3) Apply the partner or warehouse cutoff if any
-        res["date_planned"] = self._get_date_planned_with_cutoff_time(
-            res["date_planned"]
+        warehouse = self.order_id.warehouse_id
+        partner = self.order_id.partner_id
+        __, security_lead, workload = self._get_delays()
+        worload_days = self._delay_to_days(workload)
+        # Find the first date in the WH calendar by going back in the past, if any.
+        res["date_planned"] = self._deduce_workload_and_security_lead(
+            res["date_deadline"], partner, warehouse, worload_days, security_lead
         )
         return res
 
     def _prepare_procurement_values_no_commitment_date(self, res):
         """Set 'date_planned' and 'date_deadline' if no 'commitment_date'."""
-        res["date_planned"] = (
+        customer_lead, security_lead, workload = self._get_delays()
+        workload_days = self._delay_to_days(workload)
+        date_planned = (
             self.order_id.date_order
-            + timedelta(days=self.customer_lead or 0.0)
-            - timedelta(days=self.order_id.company_id.security_lead)
-        )
-        res["date_planned"] = self._get_date_planned_with_cutoff_time(
-            res["date_planned"]
-        )
-        res["date_planned"] = self._get_date_planned_with_warehouse_calendar(
-            res["date_planned"]
+            + +timedelta(days=customer_lead or 0.0)
+            - timedelta(days=security_lead)
         )
-        res["date_deadline"] = self._get_date_deadline_with_delivery_window(
-            res["date_planned"]
-        )
-        res["date_planned"] = self._get_date_planned_from_date_deadline(
-            res["date_deadline"]
+        partner = self.order_id.partner_shipping_id
+        warehouse = self.order_id.warehouse_id
+        calendar = warehouse.calendar_id
+        date_planned = self._apply_cutoff(date_planned, partner, warehouse)
+        date_planned = self._apply_workload(date_planned, workload_days, calendar)
+        date_deadline = self._apply_delivery_window(date_planned, partner, calendar)
+        date_planned = self._deduce_workload_and_security_lead(
+            date_deadline, partner, warehouse, workload_days, security_lead
         )
+        res.update({"date_deadline": date_deadline, "date_planned": date_planned})
         return res
 
-    def _get_date_planned_with_cutoff_time(self, date_planned, keep_same_day=False):
-        """Apply the cut-off time on a planned date
-
-        The cut-off configuration is taken on the partner if set, otherwise
-        on the warehouse.
+    @api.model
+    def _deduce_workload_and_security_lead(
+        self, date, partner, warehouse, days, security_lead
+    ):
+        """Return the 'date_planned' from the 'date_deadline'.
```

Comment:
Could we rename the `date` parameter to `date_deadline`? The docstring says we expected a date deadline as parameter but better to name this parameter accordingly as the name of the method doesn't tell from what it deduce the workload and security lead.
Or if you want to be agnostic about the kind of date we could process with this method, update the docstring.